### PR TITLE
available vs generated data

### PR DESCRIPTION
**Pull request by leclairm**

### CODE REVIEW

File: tests/files/configs/test_config_large.yml [hunk 1]
```diff
@@ -133,43 +133,45 @@ tasks:
...
-      type: file
-      src: postout
-  - stored_data_1:
-      type: file
-      src: stored_data
-  - stored_data_2:
-      type: file
-      src: stored_data
+  available:
+    - preproc_output:
```
Comment (GeigerJ2) [inline tests/files/configs/test_config_large.yml:137]: Isn't this of type `generated`, as it's the output of the `preproc` task, which is part of the workflow?

**Review by GeigerJ2 — APPROVED** [review]
Thank you for the PR, @leclairm! Looks fine to me. Just added one minor comment.

In addition, I think we should still have a way to denote if a data entry lives on the local machine, irrespective if it's `available` or `generated`. It could be via `type='remote'`, even though, one then loses the `file`/`dir` specification for the type. However, as I mentioned in my comment in issue #28, this actually doesn't matter for AiiDA, and I'm not sure if we really use the information somewhere else in the code. I further assume `generated` would basically _always_ be on the remote HPC?

As this isn't addressed in this PR, I'll work on it locally (and have already been), and will open another PR.

File: src/wcflow/parsing/_yaml_data_models.py [hunk 2]
```diff
@@ -133,6 +133,20 @@ def is_file_or_dir(cls, value: str) -> str:
         return value
 
 
+class ConfigAvailableData(ConfigData):
+    available: ClassVar[bool] = True
```
[the remark this answers is not in the document]
Comment (GeigerJ2) [reply src/wcflow/parsing/_yaml_data_models.py:137]: I'd steer away from using stuff that doesn't work in 3.9, though... I'm not too happy with the current requirement of needing _at least_ Python 3.12.

[the remark this answers is not in the document]
Comment (leclairm) [reply src/wcflow/parsing/_yaml_data_models.py]: I actually took a shortcut to make pydantic happy but `ClassVar` was indeed not adequate.

File: src/wcflow/parsing/_yaml_data_models.py [hunk 3]
```diff
@@ -257,7 +271,7 @@ class ConfigWorkflow(BaseModel):
     end_date: datetime
     cycles: list[ConfigCycle]
     tasks: list[ConfigTask]
-    data: list[ConfigData]
+    data: ConfigDataStore
```
[the remark this answers is not in the document]
Comment (leclairm) [reply src/wcflow/parsing/_yaml_data_models.py]: Agreed for the naming convention

File: tests/files/configs/test_config_large.yml (hunk 1 above)
Comment (leclairm) [reply tests/files/configs/test_config_large.yml:137]: In theory yes. In the example, for some reason it's not generated by a task. We'll improve the examples later I guess.

**Comment by leclairm** [comment]
> Thank you for the PR, @leclairm! Looks fine to me. Just added one minor comment.
> 
> In addition, I think we should still have a way to denote if a data entry lives on the local machine, irrespective if it's `available` or `generated`. It could be via `type='remote'`, even though, one then loses the `file`/`dir` specification for the type. However, as I mentioned in my comment in issue #28, this actually doesn't matter for AiiDA, and I'm not sure if we really use the information somewhere else in the code. I further assume `generated` would basically _always_ be on the remote HPC?
> 
> As this isn't addressed in this PR, I'll work on it locally (and have already been), and will open another PR.

Yes, let's keep that for the separate discussion.

File: src/wcflow/parsing/_yaml_data_models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (leclairm) [reply src/wcflow/parsing/_yaml_data_models.py]: > But I don't think you need the variable as you can infer the variable value from the class type.

Yes, I did that first but didn't like it then. I prefer the attribute way which reads way better when used.

File: src/wcflow/parsing/_yaml_data_models.py [hunk 4]
```diff
@@ -133,6 +134,22 @@ def is_file_or_dir(cls, value: str) -> str:
         return value
 
 
+class ConfigAvailableData(_DataBaseModel):
+
+    available: bool = True
```
Comment (agoscinski) [inline src/wcflow/parsing/_yaml_data_models.py:139]: If you keep it here, one can overwrite it in the yml file
```yml
data:
  available:
    - preproc_output:
        available: False
        type: file
        src: $PWD/examples/files/data/file.ncdf
```
Probably one can somehow add this to the class without making it an option in the yml file, but I am not sure about this approach as you already can infer the same information with a `available = isinstance(ConfigAvailableData, data` check to create the `core.Data`.

**Review by agoscinski — CHANGES_REQUESTED** [review]
Can you also adapt the `test_config_small.yml`? Most formatting issues should be fixed automatically with `hatch fmt`.

File: src/wcflow/parsing/_yaml_data_models.py (hunk 4 above)
Comment (leclairm) [reply src/wcflow/parsing/_yaml_data_models.py:139]: > If you keep it here, one can overwrite it in the yml file
> 
> ```yaml
> data:
>   available:
>     - preproc_output:
>         available: false
>         type: file
>         src: $PWD/examples/files/data/file.ncdf
> ```
> 
> Probably one can somehow add this to the class without making it an option in the yml file, but I am not sure about this approach as you already can infer the same information with a `available = isinstance(ConfigAvailableData)` check to create the `core.Data`.

Would that work for you? I think it fixes everything
```python
class _DataBaseModel(_NamedBaseModel):
    """
    To create an instance of a data defined in a workflow file.
    """

    type: str
    src: str
    format: str | None = None

    @field_validator("type")
    @classmethod
    def is_file_or_dir(cls, value: str) -> str:
        """."""
        if value not in ["file", "dir"]:
            msg = "Must be one of 'file' or 'dir'."
            raise ValueError(msg)
        return value

    @property
    def available(self) -> bool:
        return isinstance(self, ConfigAvailableData)


class ConfigAvailableData(_DataBaseModel):

    pass


class ConfigGeneratedData(_DataBaseModel):

    pass
```

Comment (agoscinski) [reply src/wcflow/parsing/_yaml_data_models.py]: I think this is ok.

**Comment by leclairm** [comment]
> Can you also adapt the `test_config_small.yml`? Most formatting issues should be fixed automatically with `hatch fmt`.

done

**Comment by leclairm** [comment]
> You can always look up the code `FBT001` online and find more explanations. Then you find such a page https://docs.astral.sh/ruff/rules/boolean-type-hint-positional-argument/ I think the arguments on the page are fair. Would make it also enforce keyword argument usage by changing to
> 
> ```python
>     def __init__(
>         self,
>         name: str,
>         type: str,  # noqa: A002
>         src: str,
>         lag: list[Duration],
>         date: list[datetime],
>         arg_option: str | None,
>         *
>         available: bool,
>     ):
> ```
> 
> I actually thought at some point to enforce keyword argument usage to every input argument to increase readability. But that we can do another time.

did it for `avaialble`

File: src/wcflow/parsing/_yaml_data_models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (leclairm) [reply src/wcflow/parsing/_yaml_data_models.py]: Resolved with porperty

**Review by agoscinski — APPROVED** [review]
Looks good! Thank for the work. Could you squash merge with a commit message like this?
```
Introduce keywords that distinguish data that is available on init and not in yml file (#29)

We introduce the keywords `available` and `generated` in the yml file that
specify if the data exists already on initialization or is generated by the
workflow. Before that we needed to refer this from how the data is used in the
workflow. We decided to let the user specify this for better error control
handling. Implements suggestion from issue #28
```

**Pull request merged**